kernel/make_initrd.sh

```diff
@@ -34,28 +34,58 @@ mount -t proc none /proc
 mount -t sysfs none /sys
 mount -t devtmpfs none /dev
 
-sleep 1
-
-if [ -e /dev/mmcblk0p2 ]; then
-	# Mount real root.
-	mkdir -p /mnt/root
-	mount -o rw /dev/mmcblk0p2 /mnt/root
-
-	if [ -x /mnt/root/sbin/init -o -h /mnt/root/sbin/init ]; then
-		# Cleanup.
-		umount /proc
-		umount /sys
-		umount /dev
-
-		# Boot the real system.
-		exec switch_root /mnt/root /sbin/init
-	fi
-
-fi
-
-echo "Dropping to a shell."
-echo
-setsid cttyhack /bin/sh
+cmdline() {
+        local value
+        value=" $(cat /proc/cmdline) "
+        value="${value##* $1=}"
+        value="${value%% *}"
+        [ "$value" != "" ] && echo "$value"
+}
+
+realboot() {
+        echo "Rootfs: $1";
+        # Mount real root.
+        mkdir -p /mnt/root
+        mount -o rw "$1" /mnt/root
+
+        if [ -x /mnt/root/sbin/init -o -h /mnt/root/sbin/init ]; then
+                # Cleanup.
+                umount /proc
+                umount /sys
+                umount /dev
+
+                # Boot the real system.
+                exec switch_root /mnt/root /sbin/init
+        else
+                umount /mnt/root
+        fi
+}
+
+runshell() {
+        echo "Dropping to a shell."
+        echo
+        setsid cttyhack /bin/sh
+}
+
+boot() {
+        echo "Kernel params: `cat /proc/cmdline`"
+        timeout=20;
+
+        while [ "$timeout" -ge 1 ]; do
+                echo "Waiting for root system $(cmdline root), countdown : $timeout";
+                if [ -e $(cmdline root) ]; then
+                        realboot $(cmdline root);
+                fi;
+
+                timeout=$(( $timeout - 1 ));
+                sleep 1;
+        done;
+
+        # Default rootfs - sd partition 2
+        realboot /dev/mmcblk0p2;
```

Comment:
This changes the behavior a little (umounting when no init was found), but i guess everyone can mount again manually in the shell for checks.